R/compare.R

```diff
@@ -0,0 +1,274 @@
+is_binary_file <- function(path, n = 1024) {
+  # This is a pretty crude heuristic, but it seems good enough.
+  # It is actually similar to what the `diff` tool does.
+  data <- readBin(path, "raw", n)
+  as.raw(0) %in% data
+}
+
+compare_metadata <- function(target, current) {
+  # id and time will almost always differ, but not in any interesting way.
+  # files may differ (especially the hashes), but we compare the files in
+  # detail seperately.
+  exclude <- c("id", "time", "files")
+
+  target_filtered <- target[!(names(target) %in% exclude)]
+  current_filtered <- current[!(names(current) %in% exclude)]
+  if (!identical(target_filtered, current_filtered)) {
+    diffobj::diffPrint(
+      target_filtered,
+      current_filtered,
+      tar.banner = target$id,
+      cur.banner = current$id,
+      rds = FALSE,
+      mode = "unified",
+      style = list(pad = FALSE, wrap = FALSE),
+      interactive = FALSE)
+  } else {
+    NULL
+  }
+}
+
+compare_filesets <- function(target, current) {
+  files <- merge(x = target, y = current, by = "path", all = TRUE)
+  status <- ifelse(
+    is.na(files$hash.x), "added",
+    ifelse(is.na(files$hash.y), "removed",
+           ifelse(files$hash.x == files$hash.y, "unchanged", "modified")))
+  data.frame(path = files$path, status = status)
+}
+
+compare_files <- function(target, current, files, root, search_options) {
```

Comment:
I feel like a non-verbose option should be really fast, and we can do this based on the hash.  That would make this sort of comparison possible even if you did not actually have the files.

It would be nice to have some option to do this quickly and just get the summary.

Similarly, there's no need to fetch the files when printing - we could just print something like "diff unavailable as files are not stored locally" or something, which would mean we could move the calls to diffobj into the print method feels more natural to me at least